Add homematicip_cloud service set cooling home

barryvdh opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Add a service call to set_cooling_home to set the entire home in Cooling mode (or disable cooling mode to revert to heating). Only works when cooling mode is enabled.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/33753

New release of homematicip-rest-api is here: https://github.com/hahn-th/homematicip-rest-api/releases/tag/1.1.2
It only includes the new set_cooling call.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

barryvdh commented:
Test failure seems unrelated to this PR. Tests run locally.

Installed repository in custom_modules. Service is available and can be called with true/false and with/without AP, similar to other services:
<img width="1138" alt="image" src="https://github.com/user-attachments/assets/1ba014d6-f510-45ec-9fda-77cd541e1321">

Doc PR: https://github.com/home-assistant/home-assistant.io/pull/33753

barryvdh committed:
[homematicip_cloud] Add service to set cooling mode

hahn-th commented on tests/components/homematicip_cloud/test_climate.py:
```diff
@@ -633,6 +633,33 @@ async def test_hmip_climate_services(
     # There is no further call on connection.
     assert len(home._connection.mock_calls) == 10
 
+    await hass.services.async_call(
```
Could you add the tests to a new test case? It's nicer, if test cases are smaller.

barryvdh replied:
I can do that, bit just followed the other tests

hahn-th commented on tests/components/homematicip_cloud/test_climate.py:
```diff
@@ -633,6 +633,33 @@ async def test_hmip_climate_services(
+    assert home.mock_calls[-1][1] == (False,)
+    assert len(home._connection.mock_calls) == 11
+
+    await hass.services.async_call("homematicip_cloud", "set_cooling", blocking=True)
```
I don't understand, why the "true" value for cooling is missing?

barryvdh replied:
The cooling value is optional. Default is True

barryvdh replied:
Fixed

barryvdh replied:
Added an extra test for true/false/no value

barryvdh committed:
Create seperate test for cooling

hahn-th commented:
Here is the changelog for bumping upstream lib from 1.1.1 to 1.1.2:
https://github.com/hahn-th/homematicip-rest-api/blob/master/CHANGELOG.md

gjohansson-ST commented on homeassistant/components/homematicip_cloud/services.py:
```diff
@@ -42,6 +43,7 @@
 SERVICE_DUMP_HAP_CONFIG = "dump_hap_config"
 SERVICE_RESET_ENERGY_COUNTER = "reset_energy_counter"
 SERVICE_SET_ACTIVE_CLIMATE_PROFILE = "set_active_climate_profile"
+SERVICE_SET_COOLING = "set_cooling"
```
Should we call it `set_home_cooling_mode` or something to make it a bit more specific what it is (+ update the wording in `strings.json`)?

gjohansson-ST replied:
The question is more related to what's it called in the app so the user "recognizes" what this service is.

gjohansson-ST reviewed (commented):
For the rest this seems like it's pretty much good to go 👍

barryvdh commented:
Do I need to do anything or request a review from a core member to merge this? Not sure of the process because it says `At least 1 approving review is required by reviewers with write access.`

gjohansson-ST commented:
> Do I need to do anything or request a review from a core member to merge this? Not sure of the process because it says `At least 1 approving review is required by reviewers with write access.`

No, me or some other will come by.
There is also a documentation PR missing for the feature so please add one and attach it into the PR description.
Undraft this PR when it's ready for review (also considering the comment I just made regarding the service name)

barryvdh commented:
Tnx, see that the PR docs was lonked in the wrong place, but it was here: https://github.com/home-assistant/home-assistant.io/pull/33753
Will update the name to set_home_cooling_mode them and report back.

barryvdh committed:
Rename service to set_home_cooling_mode

barryvdh committed:
Merge branch 'dev' into feat-homematicip_cloud-set-cooling

barryvdh commented:
I've updated the service/action and updated the docs to reflect this change. Tested this code on my own Homeassistant/Homematic installation and seems to work still.

gjohansson-ST commented on homeassistant/components/homematicip_cloud/services.py:
```diff
@@ -324,6 +344,18 @@ async def _async_reset_energy_counter(hass: HomeAssistant, service: ServiceCall)
+    cooling = service.data[ATTR_COOLING]
+
+    if hapid := service.data.get(ATTR_ACCESSPOINT_ID):
+        if home := _get_home(hass, hapid):
```
Should we raise `ServiceValidationError` here if it can't find the "home"?
Right now if an incorrect id is provided nothing happens. With `ServiceValidationError` we could give a good response to the user.

barryvdh replied:
That should probably happen in the `_get_home` function then? Currently only a log entry is added there.

>     _LOGGER.info("No matching access point found for access point id %s", hapid)

I could change that, but that would mean I would change the existing behaviour for the rest of the services. Or do you want me add that specifically for this new service?

barryvdh replied:
Should I just replace that line with 
```
raise ServiceValidationError(
    translation_domain=DOMAIN,
    translation_key="accesspoint_not_found",
    translation_placeholders={"id": hapid},
)
```
And add the exception to the strings. Or do you want me to create a new PR for this (as it changes the rest of the services also)

gjohansson-ST replied:
Yes, I think it makes sense to change the logging to using `ServiceValidationError`. Then we will also fix this for all action calls at the same time.
Don't forget to update the test.

barryvdh replied:
I added the exception + test

gjohansson-ST commented on homeassistant/components/homematicip_cloud/services.py:
```diff
@@ -324,6 +344,18 @@ async def _async_reset_energy_counter(hass: HomeAssistant, service: ServiceCall)
+        if home := _get_home(hass, hapid):
+            await home.set_cooling(cooling)
+    else:
+        for hap in hass.data[HMIPC_DOMAIN].values():
```
Does this mean we would set cooling mode for all "homes" if there is several? Is that a wanted behavior?

barryvdh replied:
If an accesspoint ID is given, it only set the cooling mode for that 'home', otherwise for all homes. But you would generally only have 1 home, so the default action doesn't require the access point ID. This is similar to the existing _async_deactivate_vacation

gjohansson-ST reviewed (changes requested):
Some final remarks/questions 👍

barryvdh committed:
Raise exception when accesspoint not found

gjohansson-ST commented:
Test failure not related so will merge this PR

gjohansson-ST reviewed (approved):
Thanks @barryvdh 👍
